pkg/destroy/azure/azure.go

```diff
@@ -92,26 +93,110 @@ func New(logger logrus.FieldLogger, metadata *types.ClusterMetadata) (providers.
 // Run is the entrypoint to start the uninstall process.
 func (o *ClusterUninstaller) Run() error {
 	o.configureClients()
+
 	group := o.InfraID + "-rg"
-	o.Logger.Debug("deleting public records")
-	if err := deletePublicRecords(context.TODO(), o.zonesClient, o.recordsClient, o.privateZonesClient, o.privateRecordSetsClient, o.Logger, group); err != nil {
-		o.Logger.Debug(err)
-		return errors.Wrap(err, "failed to delete public DNS records")
+	returnChannel := make(chan deleteResponse)
+
+	deletePublicRecordsFunc := func(name string, channel chan deleteResponse) {
+		o.Logger.Debugf("deleteing %s", name)
+		response := deleteResponse{name: name, err: nil}
+		response.err = deletePublicRecords(context.TODO(), o.zonesClient, o.recordsClient, o.privateZonesClient, o.privateRecordSetsClient, o.Logger, group)
+		if response.err != nil {
+			o.Logger.Debug(response.err)
+			response.err = errors.Wrap(response.err, "failed to delete public DNS records")
+		}
+		channel <- response
+	}
+	deleteResourceGroupFunc := func(name string, channel chan deleteResponse) {
+		o.Logger.Debugf("deleteing %s", name)
+		response := deleteResponse{name: name, err: nil}
+		response.err = deleteResourceGroup(context.TODO(), o.resourceGroupsClient, o.Logger, group)
+		if response.err != nil {
+			o.Logger.Debug(response.err)
+			response.err = errors.Wrap(response.err, "failed to delete resource group")
+		}
+		channel <- response
 	}
-	o.Logger.Debug("deleting resource group")
-	if err := deleteResourceGroup(context.TODO(), o.resourceGroupsClient, o.Logger, group); err != nil {
-		o.Logger.Debug(err)
-		return errors.Wrap(err, "failed to delete resource group")
+	deleteApplicationRegistrationsFunc := func(name string, channel chan deleteResponse) {
+		o.Logger.Debugf("deleteing %s", name)
+		response := deleteResponse{name: name, err: nil}
+		response.err = deleteApplicationRegistrations(context.TODO(), o.applicationsClient, o.serviceprincipalsClient, o.Logger, o.InfraID)
+		if response.err != nil {
+			o.Logger.Debug(response.err)
+			response.err = errors.Wrap(response.err, "failed to delete application registrations and their service principals")
+		}
+		channel <- response
+	}
+
+	deleteJobs := []deleteJob{
+		{name: "public records", deleteFunc: deletePublicRecordsFunc},
+		{name: "resource group", deleteFunc: deleteResourceGroupFunc},
+		{name: "application registrations", deleteFunc: deleteApplicationRegistrationsFunc},
```

Comment:
is it useful to create these wrappers for 3 functions? Can we for now just call these separately to keep things simple for now?

Reply:
Nuked.